File: apps/tabtin-electron/src/renderer/src/components/chat/cards/imageGeneratingProgress.ts
/**
 * 对话生图假进度：ease-out 指数逼近，封顶 92%；工具终态再冲到 100%。
 * 不接 Seedream stream——前端体感进度，非真实推理百分比。
 *
 * 返回连续浮点（不 round），供 rAF / scaleX 每帧平滑推进。
 */

export const DEFAULT_IMAGE_GENERATING_TAU_MS = 18_000

export function computeImageGeneratingProgress(params: {
  elapsedMs: number
  tauMs?: number
  done: boolean
}): number {
  if (params.done) return 100
  const tauMs = params.tauMs ?? DEFAULT_IMAGE_GENERATING_TAU_MS
  if (!(tauMs > 0) || !(params.elapsedMs >= 0)) {
    return 0
  }
  const raw = 100 * (1 - Math.exp(-params.elapsedMs / tauMs))
  return Math.min(92, raw)
}

File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/features/conversation/ImageGeneratingProgress.kt
package com.tabtin.mobile.features.conversation

import kotlin.math.exp
import kotlin.math.min

/**
 * 对话生图假进度：ease-out 指数逼近，封顶 92%；工具终态再冲到 100%。
 * 对齐 Electron `imageGeneratingProgress.ts`——连续浮点，供 scaleX 每帧平滑推进。
 */
public object ImageGeneratingProgress {
    public const val DEFAULT_TAU_MS: Long = 18_000L

    @JvmStatic
    public fun compute(
        elapsedMs: Long,
        tauMs: Long = DEFAULT_TAU_MS,
        done: Boolean,
    ): Float {
        if (done) return 100f
        if (tauMs <= 0L || elapsedMs < 0L) return 0f
        val raw = 100.0 * (1.0 - exp(-elapsedMs.toDouble() / tauMs.toDouble()))
        return min(92.0, raw).toFloat()
    }
}
